### PR TITLE
Damian/redux

### LINKED ISSUE #66: Centralize shared state
<b>Problem</b>:
As the application grows, certain pieces of state within the app now need to be shared across larger subsections of the app. As a result, the state is passed around via props and in some instances with callback functions. This leaves some of the functionality in an inconsistent state. 

<b>Proposed solution</b>
The state such as `conversations`, `contacts`, `socket` etc, for example, should be removed from their associated components (`Conversations` and `Contacts`) and be placed within a centralized store (possibly a redux store), so they can be updated; for example in the `Chats` component when a user adds a new participant to a conversation. The changes would then be reflected in real-time in the necessary locations.

### PR DESCRIPTION
**Pull request by D-Garnes**
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [x] [Searched](https://github.com/BoopChat/boop/pulls) the bugtracker for similar pull requests
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] Covered the code with tests that prove my fix is effective or that my feature works (note that PRs
without tests will be REJECTED)
- [ ] New and existing unit tests pass locally with my changes

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce.
Please also list any relevant details for your test configuration

---

### What is the purpose of your *pull request*?
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Improvement
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Changes made:
- A single instance of the socket can now be access in any component using `useContext` and passed to controller functions    as parameters where needed. The conversations controller was changed to reflect this but not the others.
- Created a socket controller which initializes the listeners for the socket. ( only the listeners for conversations have been added so far). 
- Conversations now have global state in the redux store.
- Most of the conversation state is consistent now. The closing of the chat window when the user leaves the conversation needs to be added. The `ShowChat` state and `setShowChat` reducer in the redux store was created to hanldle this. There where some inconsistencies with the closing of the chat and the success alert box. @jChris you are more familiar with how the `alertDialog` components works, so I leave that to you. 
- Moved `userSlice` from the login directory to the `redux-store` directory.

closes #66 

## Screenshots (if appropriate):

### CODE REVIEW

File: server/bin/www [hunk 1]
```diff
@@ -116,3 +116,18 @@ function getLoginService () {
...
+  const services = ["GOOGLE_CLIENT", "FACEBOOK_CLIENT", "TWITTER_CLIENT"];
+  const configuredServices = {};
+  for (const service of services) {
+    const providerName = service.split("_")[0].toLowerCase();
+    if (process.env[`${service}_SECRET`] && process.env[`${service}_ID`])
+      configuredServices[providerName] = true;
+  }
+  
+  return configuredServices;
+}
```
Comment (krazeelazy) [inline server/bin/www:133]: This is a duplicate of the previous few lines

File: client/src/components/messenger/Conversations.js [hunk 2]
```diff
@@ -205,21 +199,23 @@ const Conversations = ({ selectConversation, socket }) => {
...
-                        key={i}
-                        onClick={() => selectConversation(chat.id, chat.title, chat.participants)}
-                    />
-                ))}
+                {
+                    useSelector((state) => state.conversations.conversations)?.filter(filterConversations)
+                        .sort(sortConversations).map((chat, i) => (
+                            <ConversationItem
+                                name={chat.title}
+                                img={chat.imgUrl}
```
Comment (krazeelazy) [inline client/src/components/messenger/Conversations.js:211]: This should be reverted to `img={chat.imgUrl ?? "https://picsum.photos/400?id=" + chat.title}` so the appropriate image is shown as the chat image id imageUrl doesn't exist

**Review by krazeelazy — APPROVED** [review]
Looks good

**Review by JChris246 — APPROVED** [review]
👍

**Pull request merged**